File: 병준/BOJ_10799.java
package algorithm;
/*
문제 : 쇠 막대기 (10799)
 */
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Stack;

public class BOJ_10799 {
    public static void main(String[] args) throws IOException {
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        String inData = reader.readLine();
        String[] data = inData.split("");
        Stack<Character> stack = new Stack<>();
        int answer = 0;
        for (int i = 0; i < data.length; i++) {
            if (data[i].equals("(")) stack.push('(');  // 여는 괄호는 스택에 추가
            else if (!stack.isEmpty() && data[i].equals(")")) {
                stack.pop();  // 닫는 괄호를 만나면 가장 최근의 여는 괄호를 스택에서 제거
                // 레이저를 감지합니다. 이전 문자가 여는 괄호인 경우, 레이저로 간주
                if (data[i - 1].equals("(")) {
                    answer += stack.size();
                    // 레이저에 의해 잘린 막대의 조각 수를 추가합니다. 스택의 크기는 현재 남아있는 막대의 수를 의미합니다.
                } else {
                    answer += 1;
                    // 레이저가 아닌 경우, 막대의 끝을 의미하므로 조각 수를 1만큼 증가시킵니다.
                }
            }
            System.out.print(answer);

        }
    }
}

File: 혜민/BOJ_10799.py
#괄호 입력
array = list(input())

stack = []
count = 0
for x in array:   #입력리스트에 있는 괄호를 하나씩 확인 
    if x == "(":    
        stack.append(x)  #x가 왼쪽 괄호면 스택에 append
        before_x = x     #이전의 x값과 항상 비교해야하니까 before_x 변수에 현재 변수값 업데이트
    else:              
        if x == before_x:    #x가 오른쪽 괄호이고 이전과 같다면, 끝조각용이므로
            stack.pop()      #pop하고, 남은 끝조각 count +1
            count+=1
            before_x = x
        else:                #이전과 다르다면, 레이저용이므로 
            stack.pop()
            count+=len(stack)  #pop하고, count +현재 스택의 사이즈 
            before_x = x   
            
print(count)
